File: backend/src/pomelo_orbit/interfaces/api/credential.py
"""
凭据管理 API
"""

import math
from typing import Annotated

from fastapi import APIRouter, Depends, Query, status

from pomelo_orbit.application.credential_service import CredentialService
from pomelo_orbit.application.di import get_credential_service
from pomelo_orbit.interfaces.api.auth import get_current_user
from pomelo_orbit.interfaces.api.schemas import (
    CredentialCreateReq,
    CredentialDetailResp,
    CredentialResp,
    CredentialUpdateReq,
    PaginatedResp,
)

router = APIRouter(prefix="/credential", tags=["credential"])


@router.get("", response_model=PaginatedResp[CredentialResp])
def list_credentials(
    credential_service: Annotated[CredentialService, Depends(get_credential_service)],
    _current_user=Depends(get_current_user),
    page: Annotated[int, Query(ge=1)] = 1,
    per_page: Annotated[int, Query(ge=1, le=100)] = 10,
    search: Annotated[str | None, Query()] = None,
) -> PaginatedResp[CredentialResp]:
    """列出所有凭据（不返回值）"""
    credentials, total = credential_service.list_credentials(page, per_page, search)
    return PaginatedResp(
        items=[CredentialResp.model_validate(c) for c in credentials],
        total=total,
        page=page,
        per_page=per_page,
        pages=math.ceil(total / per_page) if total > 0 else 1,
    )


@router.get("/{credential_id}", response_model=CredentialDetailResp)
def get_credential(
    credential_id: str,
    credential_service: Annotated[CredentialService, Depends(get_credential_service)],
    _current_user=Depends(get_current_user),
) -> CredentialDetailResp:
    """获取凭据详情"""
    credential = credential_service.get_credential(credential_id)
    return CredentialDetailResp(
        id=credential.id,
        application_id=credential.application_id,
        name=credential.name,
        type=credential.type,
        value=credential_service.decrypt_credential_value(credential),
        extra_data=credential.extra_data,
        created_at=credential.created_at,
    )


@router.post("", response_model=CredentialResp, status_code=status.HTTP_201_CREATED)
def create_credential(
    data: CredentialCreateReq,
    credential_service: Annotated[CredentialService, Depends(get_credential_service)],
    _current_user=Depends(get_current_user),
) -> CredentialResp:
    """创建凭据"""
    credential = credential_service.create_credential(
        application_id=data.application_id,
        name=data.name,
        credential_type=data.type,
        value=data.value,
        extra_data=data.extra_data,
    )
    return CredentialResp.model_validate(credential)


@router.put("/{credential_id}", response_model=CredentialResp)
def update_credential(
    credential_id: str,
    data: CredentialUpdateReq,
    credential_service: Annotated[CredentialService, Depends(get_credential_service)],
    _current_user=Depends(get_current_user),
) -> CredentialResp:
    """更新凭据"""
    credential = credential_service.update_credential(
        credential_id=credential_id,
        name=data.name,
        value=data.value,
        extra_data=data.extra_data,
    )
    return CredentialResp.model_validate(credential)


@router.delete("/{credential_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_credential(
    credential_id: str,
    credential_service: Annotated[CredentialService, Depends(get_credential_service)],
    _current_user=Depends(get_current_user),
):
    """删除凭据"""
    credential_service.delete_credential(credential_id)

File: internal/transport/http/handler/ci/credential.go
package cihandler

import (
	pomeloorbit "gitee.com/leoninew/PomeloOrbit-go/internal/gen/proto/orbit/v1"
	transportcodec "gitee.com/leoninew/PomeloOrbit-go/internal/transport/http/codec"
	"github.com/gin-gonic/gin"
	"net/http"

	"gitee.com/leoninew/PomeloOrbit-go/internal/repository/model"
	cisvc "gitee.com/leoninew/PomeloOrbit-go/internal/service/ci"
	transportresponse "gitee.com/leoninew/PomeloOrbit-go/internal/transport/http/response"
)

func (h Handler) RegisterCredentialRoutes(r router) {
	r.GET("/api/ci/credential", h.listCredentials)
	r.POST("/api/ci/credential", h.createCredential)
	r.POST("/api/ci/credential/import", h.importCredential)
	r.GET("/api/ci/credential/:credential_id", h.getCredential)
	r.PUT("/api/ci/credential/:credential_id", h.updateCredential)
	r.DELETE("/api/ci/credential/:credential_id", h.deleteCredential)
	r.GET("/api/ci/credential/:credential_id/export", h.exportCredential)
}

func (h Handler) listCredentials(c *gin.Context) {
	current, ok := h.authenticator.CurrentUser(c)
	if !ok {
		return
	}
	page := transportresponse.QueryInt(c.Request.URL.Query().Get("page"), 1)
	perPage := transportresponse.QueryInt(c.Request.URL.Query().Get("per_page"), 20)
	items, err := h.service.ListCredentials(c.Request.Context(), current.Id, c.Request.URL.Query().Get("project_id"), page, perPage, c.Request.URL.Query().Get("search"))
	if err != nil {
		h.writeError(c, err)
		return
	}
	resp := mapPage(items, credentialResponse)
	c.Render(http.StatusOK, transportcodec.ProtoJSON{Message: &pomeloorbit.CredentialPaginatedResp{Items: transportresponse.Ptrs(resp.Items), Total: int32(resp.Total), Page: int32(resp.Page), PerPage: int32(resp.PerPage), Pages: int32(transportresponse.PageCount(resp.Total, resp.PerPage))}})
}

func (h Handler) createCredential(c *gin.Context) {
	current, ok := h.authenticator.CurrentUser(c)
	if !ok {
		return
	}
	var req pomeloorbit.CredentialCreateReq
	if err := transportresponse.DecodeJSON(c, &req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"detail": "Invalid JSON body"})
		return
	}
	created, err := h.service.CreateCredential(c.Request.Context(), current.Id, cisvc.CredentialCreateInput{ProjectId: c.Request.URL.Query().Get("project_id"), Name: req.Name, Type: req.Type, Data: req.Data})
	if err != nil {
		h.writeError(c, err)
		return
	}
	resp := credentialResponse(created)
	c.Render(http.StatusCreated, transportcodec.ProtoJSON{Message: &resp})
}

func (h Handler) importCredential(c *gin.Context) {
	current, ok := h.authenticator.CurrentUser(c)
	if !ok {
		return
	}
	var req pomeloorbit.CredentialImportReq
	if err := transportresponse.DecodeJSON(c, &req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"detail": "Invalid JSON body"})
		return
	}
	if req.Version == "" {
		req.Version = cisvc.CredentialExportVersion
	}
	created, err := h.service.ImportCredential(c.Request.Context(), current.Id, cisvc.CredentialCreateInput{ProjectId: c.Request.URL.Query().Get("project_id"), Name: req.Name, Type: req.Type, Data: req.Data})
	if err != nil {
		h.writeError(c, err)
		return
	}
	resp := credentialResponse(created)
	c.Render(http.StatusCreated, transportcodec.ProtoJSON{Message: &resp})
}

func (h Handler) getCredential(c *gin.Context) {
	current, ok := h.authenticator.CurrentUser(c)
	if !ok {
		return
	}
	credential, err := h.service.CredentialDetailForUser(c.Request.Context(), current.Id, c.Param("credential_id"))
	if err != nil {
		h.writeError(c, err)
		return
	}
	resp := credentialDetailResponse(credential)
	c.Render(http.StatusOK, transportcodec.ProtoJSON{Message: &resp})
}

func (h Handler) updateCredential(c *gin.Context) {
	current, ok := h.authenticator.CurrentUser(c)
	if !ok {
		return
	}
	var req pomeloorbit.CredentialUpdateReq
	if err := transportresponse.DecodeJSON(c, &req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"detail": "Invalid JSON body"})
		return
	}
	updated, err := h.service.UpdateCredential(c.Request.Context(), current.Id, c.Param("credential_id"), cisvc.CredentialUpdateInput{Name: req.Name, Data: req.Data})
	if err != nil {
		h.writeError(c, err)
		return
	}
	resp := credentialResponse(updated)
	c.Render(http.StatusOK, transportcodec.ProtoJSON{Message: &resp})
}

func (h Handler) deleteCredential(c *gin.Context) {
	current, ok := h.authenticator.CurrentUser(c)
	if !ok {
		return
	}
	if err := h.service.DeleteCredential(c.Request.Context(), current.Id, c.Param("credential_id")); err != nil {
		h.writeError(c, err)
		return
	}
	c.Status(http.StatusNoContent)
}

func (h Handler) exportCredential(c *gin.Context) {
	current, ok := h.authenticator.CurrentUser(c)
	if !ok {
		return
	}
	exported, err := h.service.ExportCredential(c.Request.Context(), current.Id, c.Param("credential_id"))
	if err != nil {
		h.writeError(c, err)
		return
	}
	c.Render(http.StatusOK, transportcodec.ProtoJSON{Message: &pomeloorbit.CredentialExportResp{Version: exported.Version, Name: exported.Name, Type: exported.Type, Data: exported.Data}})
}

func credentialResponse(item model.Credential) pomeloorbit.CredentialResp {
	return pomeloorbit.CredentialResp{Id: item.Id, Name: item.Name, Type: item.Type, CreatedAt: transportresponse.FormatTime(item.CreatedAt)}
}

func credentialDetailResponse(item cisvc.CredentialDetail) pomeloorbit.CredentialDetailResp {
	credential := credentialResponse(item.Credential)
	return pomeloorbit.CredentialDetailResp{Id: credential.Id, Name: credential.Name, Type: credential.Type, Data: item.Data, CreatedAt: credential.CreatedAt}
}
